[WIP] Implement `Encoding` in rust

b-n opened the pull request:
By extracting out `Encoding` into it's own module, we can start to rely on some of the implementations in `spinoso-string` to provide the functionality.

Some rough todos:

- [ ] Implement instance methods
- [ ] Implement constants and their binding to the interpreter

Relates: PR #2578
Relates: #183

b-n committed:
Generate trampoline for `Encoding` with NotImplemented.

b-n committed:
Manage `default_` encoding states in ruby.

Easier to maintain there, rather than implementing some sort of state engine in
rust.

b-n committed:
Plumb class methods into the trampoline

At first I thought of putting some sort of state machine into the
`mruby.rs`, but realised soon after that this makes more sense in the
trampoline since we're going to be utilizing values from `spinoso-string`

b-n committed:
Add missing parameter for `replicate` call.

b-n committed:
Implement `Encoding#name` and `Encoding#to_s`

b-n committed:
Implement `Encoding#names`

b-n committed:
Implement `Encoding#inspect`

b-n committed:
Implement `Encoding#ascii_compatible?`, `Encoding#dummy?`

b-n committed:
Implement all available `Encoding` constants from the Encoding builder

b-n committed:
Attempt number 1 with encoding and registries

Learning:
- This half works
- It pushes the allocationg of the encoding into the registry, but it's
  still backed by the extn types which feels wrong
- The encoding list doesn't actually contain all the constant registered
  encodings (we need to go the class registry spec route, I now
  understand)
- I'm slowly learning what mrb_data_type and how these things are
  hanging together.

Next steps:
- Implement a Spec type. This is the thing we want to allocate, and
  pretend is a real object. When we list, we list the spec's. Although
  Encoding objects look like they are separate because of their names,
  they are just backed by the same object on the other side.

b-n committed:
Pulling parts of Boxing out to do the assignment directly.

Most of this is copy and pasting from boxing.rs in an attempt to
directly access the encoding value from an RValue. I'm not 100% sure of
what I'm doing and if this is correct, thus this commit existing for
advice

lopopolo commented on artichoke-backend/src/convert.rs:
```diff
@@ -13,7 +13,7 @@ use crate::Artichoke;
 mod array;
 mod boolean;
-mod boxing;
+pub(crate) mod boxing;
```
this shouldn't need to be made pub. all of the types are `pub use`'d below

lopopolo commented on artichoke-backend/src/encoding_registry.rs:
```diff
@@ -0,0 +1,160 @@
+        static ESCAPABLE_CONSTANT_CHARS: OnceLock<Regex> = OnceLock::new();
+        let escapable_const_char_regex = ESCAPABLE_CONSTANT_CHARS.get_or_init(|| {
+            // Match any `-` and `.` chars. These will be replaced with `_`
+            Regex::new(r"[-\.]+").unwrap()
```
regex is really heavy here. We should be able to use `ByteSlice::replace` from `bstr`: https://docs.rs/bstr/latest/bstr/trait.ByteSlice.html#method.replace

b-n replied:
Ah nice, that is indeed much better.

I was thinking of whether there is another way to create bare symbols. e.g.

```ruby
irb(main):001:0> constants = Encoding.constants
=> 
[:ISO8859_14,                                                         
...                                                                   
irb(main):002:0> Encoding::UTF_8.replicate("Super Special Encoding")
=> #<Encoding:Super Special Encoding>
irb(main):003:0> Encoding.constants - constants
=> [:Super_Special_Encoding, :SUPER_SPECIAL_ENCODING]
irb(main):004:0> Encoding::Super_Special_Encoding
=> #<Encoding:Super Special Encoding>
irb(main):005:0> Encoding::Super_Special_Encoding.name
=> "Super Special Encoding"
```

And check this out:

```ruby
irb(main):007:0> Encoding::UTF_8.replicate("\x7F")
=> #<Encoding:>
irb(main):008:0> Encoding::UTF_8.replicate("\x80")
(Object doesn't support #inspect)
=>                                                                                
irb(main):009:0> _.name
=> "\x80"
irb(main):010:0> _.encoding
=> #<Encoding:US-ASCII>
irb(main):011:0> Encoding.find("\x80")
(Object doesn't support #inspect)
=>                                                               
irb(main):012:0> _.name
=> "\x80"
```

Valid names apparently need to be asciicompat as per https://github.com/artichoke/artichoke/blob/trunk/artichoke-backend/vendor/ruby/include/ruby/internal/encoding/encoding.h#L766-L793 which is used here: https://github.com/artichoke/artichoke/blob/trunk/artichoke-backend/vendor/ruby/encoding.c#L288-L301

Eventually some more to look into maybe, but:

```ruby
irb(main):001:0> constants = Encoding.constants
=> 
[:ISO8859_14,                                                          
...                                                                    
irb(main):002:0> (1..127).map(&:chr).each_slice(32).map { |c| Encoding::UTF_8.replicate(c.join) }
=> 
[#<Encoding:                                                           
                                                                       
                                                                       
 >,                                                                    
 #<Encoding:!"#$%&'()*+,-./0123456789:;<=>?@>,                         
 #<Encoding:ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`>,                         
 #<Encoding:abcdefghijklmnopqrstuvwxyz{|}~>]                           
irb(main):003:0> Encoding.constants - constants
=> [:ABCDEFGHIJKLMNOPQRSTUVWXYZ______, :ABCDEFGHIJKLMNOPQRSTUVWXYZ_____]
```

(Had to chunk it because otherwise some issue with Constants and lengths that I think it was getting mad at).

b-n replied:
Note: in the between time, going to take the easy Bstr replace approach.

lopopolo commented on artichoke-backend/src/extn/core/encoding/trampoline.rs:
```diff
@@ -0,0 +1,116 @@
+    Err(NotImplementedError::new().into())
+}
+
+#[allow(clippy::unnecessary_wraps)]
```
this allow appears to be unused

b-n replied:
Yeah, heavy WIP. If it looks dirty, it'll be cleaned prior to requesting real review :smile:

lopopolo commented on artichoke-backend/src/state/mod.rs:
```diff
@@ -32,6 +34,7 @@ pub struct State {
     pub parser: Option<parser::State>,
     pub classes: Registry<class::Spec>,
     pub modules: Registry<module::Spec>,
+    pub encodings: HashSet<encoding::Spec>,
```
I think a better design is to pass encodings around as symbols (i.e. a `u32`) both in the interpreter storage, `mrb_value` repr for `Encoding`s, and the flags of a `String` `mrb_value`.

If we end up doing that, we should use an `FxHashSet` here from `rustc_hash` which will perform much better

lopopolo reviewed (commented):
some early thoughts

b-n committed:
Attempt at changing mruby internals to support encoding.

There is no MRB_TT_ENCODING, so implementing as little as possible.

This code currently segfaults when trying to retrieve the value from
memory. It looks as though it does retrieve the pointer, but is
currently dieing when trying to inspect the innards.

b-n committed:
Artichoke only uses boxing_no.h

b-n committed:
No need for our own mrb_artichoke_encoding

b-n committed:
Add encoding type to mruby

This is needed since we are expecting immediates back. MRuby still needs
some sort of reference to encoding, so this init produces the smallest
possible class for it.

We could get around this by just adding the class type in instead, but
:shrug:

b-n committed:
Some fundamental changes to flow of EncodingRegistry

b-n committed:
Align encoding as a core type, and align flag to u8

This will make it consistent with what string expects.

b-n committed:
Implement String#Encoding now that we can unpack the encoding

b-n committed:
WIP - Align style with the other core extn traits.

I think it's possible to remove the EncodingRegistry entirely and maybe
relace with mezzaluna-type-registry, but will see.

b-n committed:
Use owned types in trait. This will be helpful for replica's

b-n committed:
Enable Replica encodings

TODO: There is no way to assign these encodings to a String since the
encoding is stored as bitflags at the start of the string.

b-n committed:
Re-order imports
